ДЗ (6 урок)

modimal opened the pull request:
Сделана)

modimal committed:
ДЗ (6 урок)

modimal committed:
ДЗ (7 урок)

Сделал лото)

bgaifullin commented on lesson07/loto.py:
```diff
@@ -0,0 +1,126 @@
+            i = 0
+            while i < 5:
+                num = random.randrange(FROM_N, TO_N)
+                if not self.is_used(num):
```
очень сложная получилась функция, для каждого случайного числа надо провести поиск в 3 массивах.
лучше было использовать random.sample

bgaifullin commented on lesson07/loto.py:
```diff
@@ -0,0 +1,126 @@
+class Ticket:
+    def __init__(self):
+        self._row1, self._row2, self._row3 = [], [], []
+        self._rows = {'row1': self._row1, 'row2': self._row2, 'row3': self._row3}
```
не понял, зачем хранить строки отдельно и еще в словаре

bgaifullin commented on lesson07/loto.py:
```diff
@@ -0,0 +1,126 @@
+    def crossing(self, num):
+        row_n = 'row' + str(self.is_used(num))
+        row = self._rows[row_n]
+        row.remove(row[row.index(num)])
```
row.remove(num), мы уже проверили что num в указанном row
можно было использовать матрицу self.rows = [[первая пятерка], [вторая пятерка], [третья пятерка]]

bgaifullin commented on lesson07/loto.py:
```diff
@@ -0,0 +1,126 @@
+        print('В этот раз выиграл компьютер.')
+        exit(0)
+
+    random_keg_i = random.randrange(len(kegs))
```
это не защищает нас от того, что один и тот же бочонок будет выбран два раза.
нужное либо random.choice 
либо перемешать бочонки используя random.shuffle и потом всегда доставать с конца списка с удалением
либо вариант который рассматривали на уроке

modimal committed:
Переделка игры лото

modimal committed:
ДЗ (8 урок)

Я изменил логику работы. Вместо того, чтобы скачивать весь файл с городами, который весит 28Мб+, я спрашиваю конкретные города и страны, и делаю запрос в openweather.

bgaifullin commented on lesson07/lotto.py:
```diff
@@ -0,0 +1,197 @@
+        winner = None
+        while winner is None:
+            random.shuffle(self._kegs)
+            randkeg = self._kegs[len(self._kegs) - 1]
```
можно было просто randkey = self._kegs.pop(), эта функция делает тоже самое что тут написано )

modimal replied:
а да, загуглил, верно)

modimal replied:
буду знать)

bgaifullin commented on lesson07/lotto.py:
```diff
@@ -0,0 +1,197 @@
+    def crossing(self, num):
+        if not self._check_num(num):
+            return False
+        fr = self._ticket._formatted.replace
```
ох, есть же массив чисел, можно наверное было использовать его для логики  проверки и зачеркивания. 
а форматирование в строку делать при каждой итерации игрового цикла.
чтобы не пришлось использовать замену текста в строке для зачеркивания и поиск по регулярному выражению для определения все номера зачеркнуты или не все.

bgaifullin commented on lesson08/openweather.py:
```diff
@@ -0,0 +1,149 @@
+                    );
+                    """)
+        except sqlite3.OperationalError:
+            pass
```
стоит все таки хотя бы в консоль напечатать ошибку.

bgaifullin commented on lesson08/openweather.py:
```diff
@@ -0,0 +1,149 @@
+            self._update_data(city)
+
+    def _insert_data(self, city):
+        with sqlite3.connect(self.dbname) as conn:
```
наверное, если это контроллер, тогда подключение стоит открывать один раз в конструкторе и далее использовать его, чтобы не приходилось постоянно открывать новое подключение.

modimal replied:
а когда его закрывать?

modimal replied:
если закроешь, придется новое открывать, верно?

modimal replied:
то есть когда игра закончена, закрывать только тогда?

bgaifullin commented on lesson08/openweather.py:
```diff
@@ -0,0 +1,149 @@
+        self._name = name.lower()
+        self._code = code
+        with open('app.id', 'r', encoding='utf-8') as f:
+            self._appid = f.readline()
```
отлично. вот это правильное решение, читаем app_id из файла.

modimal replied:
спасибо)
